File: src/ResidApp.Domain/Residents/Resident.cs
using ResidApp.Shared;

namespace ResidApp.Domain.Residents;

/// <summary>Traduce ResidentAdministrativeLocation de resident.ts.</summary>
public sealed record ResidentAdministrativeLocation(string? Building = null, string? Floor = null, string? Room = null, string? Place = null)
{
    public static readonly ResidentAdministrativeLocation Empty = new();
}

/// <summary>
/// Traduce ResidentAdministrativeIdentity de lib/domain/residents/resident.ts (identidad administrativa
/// mínima exigida por RES-01). DisplayName y BirthDate usan propiedades respaldadas por 'field' (C# 14)
/// para validación compacta de una sola propiedad, tal como se pidió explícitamente: no vacío/recortado
/// y fecha de nacimiento no futura, replicando las comprobaciones de
/// db/repositories/resident-repository.ts (assertInput).
/// </summary>
public sealed class Resident
{
    public required ResidentId Id { get; init; }
    public required CenterId CenterId { get; init; }
    public required UnitId UnitId { get; init; }

    public required string DisplayName
    {
        get;
        init => field = !string.IsNullOrWhiteSpace(value)
            ? value.Trim()
            : throw new DomainValidationException("RESIDENT_CREATE_INPUT_INVALID");
    }

    public required DateOnly BirthDate
    {
        get;
        init => field = value <= DateOnly.FromDateTime(DateTime.UtcNow)
            ? value
            : throw new DomainValidationException("RESIDENT_CREATE_INPUT_INVALID");
    }

    public required DocumentedSexCode DocumentedSex { get; init; }
    public required ResidentStatus Status { get; init; }
    public ResidentAdministrativeLocation Location { get; init; } = ResidentAdministrativeLocation.Empty;
}

File: docs/legado-cloudflare/lib/domain/residents/resident.ts
import type { CenterId, ResidentId, UnitId } from "../shared/identifiers.ts";

export type ResidentAdministrativeLocation = Readonly<{
  building?: string;
  floor?: string;
  room?: string;
  place?: string;
}>;

export const DOCUMENTED_SEX_CODES = ["male", "female", "other", "unknown"] as const;

export type DocumentedSexCode = (typeof DOCUMENTED_SEX_CODES)[number];

export const RESIDENT_STATUSES = ["ACTIVO", "INACTIVO"] as const;

export type ResidentStatus = (typeof RESIDENT_STATUSES)[number];

/**
 * Identidad administrativa mínima exigida por RES-01.
 * El sexo procede de documentación administrativa y nunca se infiere ni admite texto libre.
 */
export type ResidentAdministrativeIdentity = Readonly<{
  id: ResidentId;
  centerId: CenterId;
  unitId: UnitId;
  name: string;
  birthDate: string;
  documentedSex: DocumentedSexCode;
  status: ResidentStatus;
  location: ResidentAdministrativeLocation;
}>;

export function isDocumentedSexCode(value: unknown): value is DocumentedSexCode {
  return (
    typeof value === "string" &&
    (DOCUMENTED_SEX_CODES as readonly string[]).includes(value)
  );
}

export type ResidentScope = Readonly<{
  centerId: CenterId;
  unitId: UnitId;
  residentId: ResidentId;
}>;
